iroh/src/endpoint.rs

```diff
@@ -1271,33 +1279,342 @@ impl Future for Connecting {
+    #[inline]
+    pub fn remote_address(&self) -> SocketAddr {
+        self.inner.remote_address()
+    }
+
+    /// The local IP address which was used when the peer established the connection.
+    ///
+    /// This can be different from the address the endpoint is bound to, in case the
+    /// endpoint is bound to a wildcard address like `0.0.0.0` or `::`.
+    ///
+    /// This will return `None` for clients, or when the platform does not expose this
+    /// information. See [`quinn::udp::RecvMeta::dst_ip`] for a list of supported
+    /// platforms.
+    #[inline]
+    pub fn local_ip(&self) -> Option<IpAddr> {
+        self.inner.local_ip()
+    }
+
+    /// Current best estimate of this connection's latency (round-trip-time).
+    #[inline]
+    pub fn rtt(&self) -> Duration {
+        self.inner.rtt()
+    }
+
+    /// Returns connection statistics.
+    #[inline]
+    pub fn stats(&self) -> ConnectionStats {
+        self.inner.stats()
+    }
+
+    /// Current state of the congestion control algorithm, for debugging purposes.
+    #[inline]
+    pub fn congestion_state(&self) -> Box<dyn quinn_proto::congestion::Controller> {
+        self.inner.congestion_state()
+    }
+
+    /// Parameters negotiated during the handshake.
+    ///
+    /// Guaranteed to return `Some` on fully established connections or after
+    /// [`Connecting::handshake_data()`] succeeds. See that method's documentations for
+    /// details on the returned value.
+    ///
+    /// [`Connection::handshake_data()`]: crate::Connecting::handshake_data
+    #[inline]
+    pub fn handshake_data(&self) -> Option<Box<dyn Any>> {
+        self.inner.handshake_data()
+    }
+
+    /// Extracts the ALPN protocol from the peer's handshake data.
+    pub fn alpn(&self) -> Option<Vec<u8>> {
+        let data = self.handshake_data()?;
+        match data.downcast::<quinn::crypto::rustls::HandshakeData>() {
+            Ok(data) => data.protocol,
+            Err(_) => None,
+        }
+    }
+
+    /// Cryptographic identity of the peer.
+    ///
+    /// The dynamic type returned is determined by the configured [`Session`]. For the
+    /// default `rustls` session, the return value can be [`downcast`] to a
+    /// <code>Vec<[rustls::pki_types::CertificateDer]></code>
+    ///
+    /// [`Session`]: quinn_proto::crypto::Session
+    /// [`downcast`]: Box::downcast
+    #[inline]
+    pub fn peer_identity(&self) -> Option<Box<dyn Any>> {
+        self.inner.peer_identity()
+    }
+
+    /// Returns the [`NodeId`] from the peer's TLS certificate.
+    ///
+    /// The [`PublicKey`] of a node is also known as a [`NodeId`].  This [`PublicKey`] is
+    /// included in the TLS certificate presented during the handshake when connecting.
+    /// This function allows you to get the [`NodeId`] of the remote node of this
+    /// connection.
+    ///
+    /// [`PublicKey`]: iroh_base::PublicKey
+    // TODO: Would be nice if this could be infallible.
```

Comment:
Are there cases where we expect this to be fallible?
Why not make this infallible by using `self.peer_identity().unwrap()`?
Is returning an error here a [boneheaded exception](https://ericlippert.com/2008/09/10/vexing-exceptions/)?

Reply:
I don't find that error classification and vocabulary particularly easy, but I do appreciate the broad categorisation attempt in it.  Which is a long winded way of saying I could have misunderstood it, but I do think it probably is a boneheaded exception.

*Today*

Today we only try to connect to remote peers that have exactly this certificate setup.  But what when we do this raw certificate stuff?  What about another future change?  Would it be a mistake to be so strict here in old clients?  Would the length check for exactly 1 become a problem?  I don't know.

So... I don't know, and I'm unsure about making such an important change in an otherwise unrelated refactoring.  I feel like it deserves it's own PR and commit in main.

So I made a separate issue for it: #3123